From 2b9e872848c5b1bc004ee272eac11cef6a9978e0 Mon Sep 17 00:00:00 2001
From: Kevin Hao <kexin.hao@windriver.com>
Date: Tue, 5 Mar 2013 11:08:02 +0800
Subject: [PATCH 469/666] powerpc/book3e: disable the secondary threads when
 kexec boot

In the embedded multi-threaded implementation, all the threads in a
core share the TLB. On the kexec boot cpu, it will invalidate all the
TLB entries except the one which is used to boot. This will also
invalidate the 1:1 TLB entry which the secondary threads in the same
core are used to spin. Then cause crash on these threads. So we should
disable all the secondary threads when doing kexec boot.

Due to the above reasons, we definitely can't choose a secondary
thread as kexec boot cpu. Since the boot_cpuid is aways thread0,
just choose it.

Signed-off-by: Kevin Hao <kexin.hao@windriver.com>
Signed-off-by: Bin Jiang <bin.jiang@windriver.com>
---
 arch/powerpc/kernel/misc_64.S     | 16 ++++++++++++++--
 arch/powerpc/platforms/85xx/smp.c |  3 +++
 2 files changed, 17 insertions(+), 2 deletions(-)

diff --git a/arch/powerpc/kernel/misc_64.S b/arch/powerpc/kernel/misc_64.S
index b0afd47..5ffb097 100644
--- a/arch/powerpc/kernel/misc_64.S
+++ b/arch/powerpc/kernel/misc_64.S
@@ -473,8 +473,20 @@ _GLOBAL(scom970_write)
  * Physical (hardware) cpu id should be in r3.
  */
 _GLOBAL(kexec_wait)
-	bl	1f
-1:	mflr	r5
+#ifdef CONFIG_PPC_BOOK3E
+BEGIN_FTR_SECTION
+	/* Disable all the secondary threads */
+	andi.	r4,r3,1
+	beq	1f
+	li	r4,2
+	mtspr	SPRN_TENC,r4
+	b	$
+1:
+END_FTR_SECTION_IFSET(CPU_FTR_SMT)
+#endif
+
+	bl	2f
+2:	mflr	r5
 	addi	r5,r5,kexec_flag-1b
 
 99:	HMT_LOW
diff --git a/arch/powerpc/platforms/85xx/smp.c b/arch/powerpc/platforms/85xx/smp.c
index c0481ca..5d10892 100644
--- a/arch/powerpc/platforms/85xx/smp.c
+++ b/arch/powerpc/platforms/85xx/smp.c
@@ -454,6 +454,9 @@ static void mpc85xx_smp_machine_kexec(struct kimage *image)
 	}
 #endif
 
+	if (cpu_has_feature(CPU_FTR_SMT))
+		set_cpus_allowed_ptr(current, cpumask_of(boot_cpuid));
+
 	default_machine_kexec(image);
 }
 #endif /* CONFIG_KEXEC_CORE */
-- 
2.7.4

